---
title: "QC Report - NextFlow RNA-Seq run"
author: "RNA-Seq NF"
date: "`r format(Sys.time(), '%d %B %Y')`"
logo: "logo_gallery.png"
output:
  epuRate::epurate:
    toc: TRUE
    number_sections: FALSE
    code_folding: "hide"
---

```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE)
library(rmarkdown)
library(epuRate)
```

```{r, code = readLines("QC_RNASeq.R"), include=F}
```

Rscript -e "rmarkdown::render('QC.Rmd')" -i ~/test/testRNA/output

# Overview - Input samples.

Overview of the input files per sample. If paired-end, both R1 and R2 files are shown.

```{r, message=F}
  ggplotly(plots$overviewFastq)
```


# TrimGalore

This overview depicts the number of input reads (R1/R2) present within the given fastq files and the total number of respective reads containing adapter sequences.

```{r, message=FALSE}
ggplotly(plots$TrimGalore)
```

# SortMeRNA

The percentage of (trimmed) reads passing the E-value threshold meaning they have a high-confidence hit against one of the ribosomal databases.

```{r, message=FALSE}
ggplotly(plots$sortMeRNA)
```

# Complexity curves.

Complexity curves depict the ..

```{r, message=FALSE}
ggplotly(plots$preseq)
```

# Flagstat

```{r, message=FALSE, warning=FALSE}
ggplotly(plots$flagstat)
```

# TIN

```{r, message=FALSE, warning=FALSE}
ggplotly(plots$TIN)
```
